template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -243,7 +247,51 @@ private Map<String, Set<ServiceComponent>> getNonGatewayServicesByHostGroup() {
                                 .filter(i -> !i.getComponent().equalsIgnoreCase("GATEWAY"))
                                 .collect(toSet())
                 ));
+    }
+
+    private Set<String> collectDependComponentsFromGivenComponenets(Map<String, String> roleDependencyMap) {
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponents = getNonGatewayServicesByHostGroup();
+        return hgToNonGwServiceComponents.values().stream()
+                .map(serviceComponents -> roleDependencyMap.get(getDependentService(serviceComponents, roleDependencyMap)))
+                .collect(toSet());
+    }
+
+    //If the component is not present in the roleDependencyMap then we will return "componentNotPresentinRoleDependencyMap"
+    // and it is mapped to "noDependentComponentFound".
+    //In this case, First if we find any dependent component hostGroup then will first do its health check,
+    // and then we will follow our old approach and also check if the stack and stopped instances are avaiable or not.
+
+    private String getDependentService(Set<ServiceComponent> serviceComponents, Map<String, String> roleDependencyMap) {
+        Set<String> serviceComponentRoles = collectComponents(serviceComponents);
+        return roleDependencyMap.keySet().stream().filter(serviceComponentRoles::contains).findFirst().orElse("componentNotPresentinRoleDependencyMap");
+    }
+
+    public List<String> getUnhealthyDependentHostGroups(StackDto stackdto, Map<String, String> roleDependencyMap) {
+            Map<String, Set<String>> nonGatewayServices = getNonGatewayComponentsByHostGroup();
+            Set<String> dependentComponent = collectDependComponentsFromGivenComponenets(roleDependencyMap);
+            Set<String> dependentHostGroupNames = nonGatewayServices.entrySet().stream()
+                    .filter(allComponentsInHostGroup ->
+                            isDependentComponentPresentInHostGroup(allComponentsInHostGroup.getValue(), dependentComponent))
+                    .map(Entry::getKey)
+                    .collect(toSet());
+            List<String> allUnHealthyHostGroupNames = stackdto.getAllAvailableInstances().stream()
+                    .filter(instanceMetaData -> !InstanceStatus.SERVICES_HEALTHY.equals(instanceMetaData.getInstanceStatus())
+                            && !InstanceStatus.STOPPED.equals(instanceMetaData.getInstanceStatus()))
```

Comment:
@sidseth would this be the resultant check for instanceStatus? Considering the hostgroup can have stopped instances